File: src/workspaceLinkService.ts
import * as vscode from 'vscode';
import { LoggingService } from './LoggingService';

const STORAGE_KEY = 'pipelinesexplorer.repoLinks.v1';

export interface RepoLinkKey {
	orgAccountId: string;
	projectId: string;
	repoKey: string;
}

/**
 * Persists a mapping from (organization, project, repository) to a local
 * workspace folder path. Used to resolve template / script references to
 * files on disk when the user double-clicks a tree item.
 */
export class WorkspaceLinkService {
	private readonly _onDidChange = new vscode.EventEmitter<void>();
	readonly onDidChange = this._onDidChange.event;

	constructor(
		private readonly context: vscode.ExtensionContext,
		private readonly logger: LoggingService,
	) {}

	private encodeKey(k: RepoLinkKey): string {
		return `${k.orgAccountId}::${k.projectId}::${k.repoKey}`;
	}

	private readAll(): Record<string, string> {
		return this.context.globalState.get<Record<string, string>>(STORAGE_KEY, {});
	}

	private async writeAll(value: Record<string, string>): Promise<void> {
		await this.context.globalState.update(STORAGE_KEY, value);
		this._onDidChange.fire();
	}

	get(key: RepoLinkKey): string | undefined {
		return this.readAll()[this.encodeKey(key)];
	}

	async set(key: RepoLinkKey, fsPath: string): Promise<void> {
		const all = this.readAll();
		all[this.encodeKey(key)] = fsPath;
		await this.writeAll(all);
		this.logger.logInfo(`Linked ${this.encodeKey(key)} -> ${fsPath}`);
	}

	async remove(key: RepoLinkKey): Promise<void> {
		const all = this.readAll();
		if (delete all[this.encodeKey(key)]) {
			await this.writeAll(all);
			this.logger.logInfo(`Unlinked ${this.encodeKey(key)}`);
		}
	}

	/** Look up by repoKey alone — useful when a template references a repo we
	 *  haven't seen as a tree node (cross-project resource reference). */
	findAnyByRepoKey(repoKey: string): string | undefined {
		const all = this.readAll();
		for (const [k, v] of Object.entries(all)) {
			if (k.endsWith(`::${repoKey}`)) {
				return v;
			}
		}
		return undefined;
	}
}

File: src/vs2026/Services/WorkspaceLinkService.cs
using System;
using System.Collections.Generic;

namespace PipelinesExplorer.VisualStudio.Services;

/// <summary>
/// Persists a mapping from <see cref="RepoLinkKey"/> to a local folder path.
/// Used to resolve template/script references to files on disk when the user
/// activates a tree item. Mirrors <c>WorkspaceLinkService</c> in the VS Code
/// client (storage key: <c>pipelinesexplorer.repoLinks.v1</c>).
/// </summary>
public sealed class WorkspaceLinkService
{
    private const string StorageKey = "pipelinesexplorer.repoLinks.v1";

    private readonly LoggingService _logger;
    private readonly JsonStateStore _store;

    public WorkspaceLinkService(LoggingService logger, JsonStateStore? store = null)
    {
        _logger = logger;
        _store = store ?? JsonStateStore.Shared;
    }

    /// <summary>Raised whenever the underlying mapping is mutated.</summary>
    public event EventHandler? Changed;

    public string? Get(RepoLinkKey key)
    {
        return ReadAll().TryGetValue(key.Encode(), out var v) ? v : null;
    }

    public void Set(RepoLinkKey key, string fsPath)
    {
        var all = ReadAll();
        all[key.Encode()] = fsPath;
        WriteAll(all);
        _logger.Info($"Linked {key.Encode()} -> {fsPath}");
    }

    public void Remove(RepoLinkKey key)
    {
        var all = ReadAll();
        if (all.Remove(key.Encode()))
        {
            WriteAll(all);
            _logger.Info($"Unlinked {key.Encode()}");
        }
    }

    /// <summary>
    /// Look up by <c>repoKey</c> alone — useful when a template references a
    /// repository we have not seen as a tree node (cross-project resource ref).
    /// </summary>
    public string? FindAnyByRepoKey(string repoKey)
    {
        var suffix = "::" + repoKey;
        foreach (var kv in ReadAll())
        {
            if (kv.Key.EndsWith(suffix, StringComparison.Ordinal))
            {
                return kv.Value;
            }
        }
        return null;
    }

    private Dictionary<string, string> ReadAll() =>
        _store.Get(StorageKey, new Dictionary<string, string>(StringComparer.Ordinal));

    private void WriteAll(Dictionary<string, string> value)
    {
        _store.Set(StorageKey, value);
        Changed?.Invoke(this, EventArgs.Empty);
    }
}
